From 1c9b044118ba5dbcc5bf868962625ee89486c8a0 Mon Sep 17 00:00:00 2001
From: Oliver Steffen <osteffen@redhat.com>
Date: Wed, 15 Oct 2025 13:17:36 -0400
Subject: [PATCH 27/27] OvmfPkg/IoMmuDxe: Fix 1M and 2M buffer handling

The IoMmu keeps a pool of pre-allocated shared buffers in various sizes
to serve requests. Usage is tracked in a bitmap.

The bitmap masks for the 1M and 2M buffer pools are incorrect, causing
the same buffers getting handed out repeatedly, causing corrupted device
accesses.

The masks needs to be kept in sync with mReservedMemRanges below.
This sets the correct values for:

 - RESERVED_MEM_BITMAP_1M_MASK = (1 << 14) = 0x4000
 - RESERVED_MEM_BITMAP_2M_MASK = (1 << 15) | (1 << 16) = 0x18000

Signed-off-by: Oliver Steffen <osteffen@redhat.com>
---
 OvmfPkg/IoMmuDxe/IoMmuBuffer.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/OvmfPkg/IoMmuDxe/IoMmuBuffer.c b/OvmfPkg/IoMmuDxe/IoMmuBuffer.c
index f8dcd5b7ec92..6b9c38cf0db2 100644
--- a/OvmfPkg/IoMmuDxe/IoMmuBuffer.c
+++ b/OvmfPkg/IoMmuDxe/IoMmuBuffer.c
@@ -23,8 +23,8 @@ extern BOOLEAN  mReservedSharedMemSupported;
 #define RESERVED_MEM_BITMAP_4K_MASK    0xf
 #define RESERVED_MEM_BITMAP_32K_MASK   0xff0
 #define RESERVED_MEM_BITMAP_128K_MASK  0x3000
-#define RESERVED_MEM_BITMAP_1M_MASK    0x40000
-#define RESERVED_MEM_BITMAP_2M_MASK    0x180000
+#define RESERVED_MEM_BITMAP_1M_MASK    0x4000
+#define RESERVED_MEM_BITMAP_2M_MASK    0x18000
 #define RESERVED_MEM_BITMAP_MASK       0x1fffff
 
 /**
-- 
2.51.0

